clippy_lints/src/unnecessary_wraps.rs

```diff
@@ -83,22 +85,43 @@ impl<'tcx> LateLintPass<'tcx> for UnnecessaryWraps {
             }
         }
 
-        let (return_type, path) = if is_type_diagnostic_item(cx, return_ty(cx, hir_id), sym::option_type) {
+        // Check if return type is Option or Result. If neither, abort.
+        let return_ty = return_ty(cx, hir_id);
+        let (return_type_label, path) = if is_type_diagnostic_item(cx, return_ty, sym::option_type) {
             ("Option", &paths::OPTION_SOME)
-        } else if is_type_diagnostic_item(cx, return_ty(cx, hir_id), sym::result_type) {
+        } else if is_type_diagnostic_item(cx, return_ty, sym::result_type) {
             ("Result", &paths::RESULT_OK)
         } else {
             return;
         };
 
+        // Take the first inner type of the Option or Result. If can't, abort.
+        let inner_ty = if_chain! {
```

Comment:
This used to be in span_lint_and_then() below. I moved it here to check if the inner type is unit.

Reply:
This is good. You could simplify this a little more and merge it with the code above like this:

```rust
let (return_type_label, path, inner_ty) = if let ty::Adt(def_id, subst) = return_ty.kind() {
    if cx.tcx.is_diagnostic_item(sym::option_type, def_id) {
		    ("Option", .., subst.type_at(0))
	} else if cx.tcx.is_diagnostic_item(sym::result_type, def_id) {
			("Result", .., subst.type_at(0))
	} else {
			return;
	}
}
```